docs/opentracing.rst

```diff
@@ -0,0 +1,179 @@
+
+
+Enable opentracing in synapse
+-----------------------------
+
+Opentracing is not enabled by default. It must be enabled in the homeserver
+config by uncommenting the config options under ``opentracing``. For example:
+
+.. code-block:: yaml
+
+  opentracing:
+    # Enable / disable tracer
+    tracer_enabled: true
+    # The list of homeservers we wish to expose our current traces to.
+    # The list is a list of regexes which are matched against the
+    # servername of the homeserver
+    homeserver_whitelist:
+      - ".*"
+
+Homeserver whitelisting
+-----------------------
+
+The homeserver whitelist is configured using regex. A list of regexes can be
+given and their union will be compared when propagating any spans through a
+carrier. Most of the whitelist checks are encapsulated in the lib's injection
+and extraction method but be aware that using custom carriers or crossing
+unchartered waters will require the enforcement of this whitelist.
+
+``logging/opentracing.py`` has a ``whitelisted_homeserver`` method which takes
+in a destination and compares it to the whitelist.
+
+============================
+Using opentracing in synapse
+============================
+
+Access to the opentracing api is mediated through the
+``logging/opentracing.py`` lib. Opentracing is encapsulated such that
+no statefull spans from opentracing are used in synapses code. This allows
+opentracing to be easily disabled in synapse and thereby have opentracing as
+an optional dependency. This does however limit the number of modifyable spans
+at any point in the code to one. From here out references to opentracing refer
+to the lib implemented in synapse.
+
+Tracing
+-------
+
+In synapse it is not possible to start a non-active span. Spans can be started
+using the ``opentracing.start_active_span`` method. This returns a scope (see
+opentracing docs) which is a context manager that needs to be entered and
+exited. This is usually done by using ``with``.
+
+.. code-block:: python
+
+   with start_active_span("operation name"):
+       # Do something we want to tracer
+
+Forgetting to enter or exit a scope will result in some mysterious grevious log
+context errors.
+
+At anytime where there is an active span ``opentracing.set_tag`` can be used to
+set a tag on the current active span.
+
+Tracing functions
+-----------------
+
+Functions can be easily traced using decorators. There is a decorator for
+'normal' function and for functions which are actually deferreds. The name of
+function becomes the operation name for the span.
+
+.. code-block:: python
+
+   # Start a span using 'normal_function' as the operation name
+   @trace
+   def normal_function(*args, **kwargs):
+       # Does all kinds of cool and expected things
+       return something_usual_and_useful
+
+   # Start a span using 'deferred_function' as the operation name
+   @trace_deferred
+   # Yes, there is a typo in the lib. I will fix this
+   def deferred_function(*args, **kwargs):
```

Comment:
should this have an `@defer.inlineCallbacks` too?

Reply:
yup